src/v/storage/tests/concat_segment_reader_test.cc

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "model/tests/random_batch.h"
+#include "storage/directories.h"
+#include "storage/log_manager.h"
+#include "storage/tests/utils/disk_log_builder.h"
+#include "test_utils/tmp_dir.h"
+
+#include <seastar/testing/thread_test_case.hh>
+#include <seastar/util/defer.hh>
+
+constexpr size_t segment_size{32_MiB};
+
+using namespace storage;
+
+size_t copy_stream(concat_segment_reader_view& cv) {
+    auto concat_str = cv.take_stream();
+    iobuf buffer{};
+    auto ostream = make_iobuf_ref_output_stream(buffer);
+    ss::copy(concat_str, ostream).get();
+
+    // We do not actually need to close this, because the stream has been moved
+    // out of this view.
+    cv.close().get();
+
+    // Closes the stream, the data source and the
+    // concat_segment_data_source_impl
+    concat_str.close().get();
+    auto sz = buffer.size_bytes();
+    ostream.close().get();
+    return sz;
+}
+
+SEASTAR_THREAD_TEST_CASE(
```

Comment:
nit: Do we ever perform reads that are not on segment boundaries? If so, it might be interesting to adjust the test to do that too by adding an offset to start_pos or end_pos.

Reply:
We do perform reads starting inside segments. For an example see `archival::get_file_range` which seeks inside segments files for given `model::offset` and then reads from batch including that offset in the segment file. Usually this happens when we have set SI to force upload data every few seconds, in that case we upload open segments and have to seek inside segment files to upload from the last uploaded offset. 

With compacted segment uploads this will become more frequent when we try to align uploads to manifest segments.

This test and a couple of other tests in this PR do test reading inside segment boundaries, here we have start_pos one batch inside the first segment and end_pos one batch before the end of the last segment. Same in `test_single_segment_read_with_bounds`.

Note that we do not start or end read inside batch boundaries.